### PR TITLE
Fix validator compatibility with Symfony 3

### LINKED ISSUE #470: ["validateOneMainCategory"] targeted by Callback constraint is not a valid callable (Symfony 3)
#### Sonata packages

```
gsonata-project/admin-bundle              3.23.0               3.23.0               The missing Symfony Admin Gen...
sonata-project/block-bundle              3.4.0                3.4.0                Symfony SonataBlockBundle
sonata-project/cache                     1.0.7                1.0.7                Cache library
sonata-project/cache-bundle              2.3.1                2.3.1                This bundle provides caching ...
sonata-project/classification-bundle     3.3.1                3.3.1                Symfony SonataClassificationB...
sonata-project/comment-bundle            3.0.0                3.0.0                Integrate the FOSCommentBundl...
sonata-project/core-bundle               3.5.1                3.5.1                Symfony SonataCoreBundle
sonata-project/datagrid-bundle           2.2.1                dev-master 0914303   Symfony SonataDatagridBundle
sonata-project/doctrine-extensions       1.0.2                1.0.2                Doctrine2 behavioral extensions
sonata-project/doctrine-orm-admin-bundle 3.1.7                3.1.7                Symfony Sonata / Integrate Do...
sonata-project/easy-extends-bundle       2.2.0                2.2.0                Symfony SonataEasyExtendsBundle
sonata-project/ecommerce                 dev-symfony3 5253d8d dev-symfony3 5253d8d E-Commerce solution provided ...
sonata-project/exporter                  1.7.1                1.7.1                Lightweight Exporter library
sonata-project/formatter-bundle          3.2.3                3.2.3                Symfony SonataFormatterBundle
sonata-project/intl-bundle               2.2.4                2.3.2                Symfony SonataIntlBundle
sonata-project/media-bundle              3.6.0                3.6.0                Symfony SonataMediaBundle
sonata-project/notification-bundle       3.2.0                3.2.0                Symfony SonataNotificationBundle
sonata-project/page-bundle               3.5.2                3.5.2                This bundle provides a Site a...
sonata-project/seo-bundle                2.2.0                2.2.0                Symfony SonataSeoBundle
sonata-project/translation-bundle        2.2.0                2.2.0                SonataTranslationBundle
sonata-project/user-bundle               dev-master a962eaa   dev-master a962eaa   Symfony SonataUserBundle
```

```
symfony/monolog-bundle     v3.1.1 v3.1.1 Symfony MonologBundle
symfony/phpunit-bridge     v3.3.9 v3.3.9 Symfony PHPUnit Bridge
symfony/polyfill-apcu      v1.5.0 v1.5.0 Symfony polyfill backporting apcu_* functions to lower PHP versions
symfony/polyfill-intl-icu  v1.5.0 v1.5.0 Symfony polyfill for intl's ICU-related data and classes
symfony/polyfill-mbstring  v1.5.0 v1.5.0 Symfony polyfill for the Mbstring extension
symfony/polyfill-php56     v1.5.0 v1.5.0 Symfony polyfill backporting some PHP 5.6+ features to lower PHP versions
symfony/polyfill-php70     v1.5.0 v1.5.0 Symfony polyfill backporting some PHP 7.0+ features to lower PHP versions
symfony/polyfill-util      v1.5.0 v1.5.0 Symfony utilities for portability of PHP codes
symfony/security-acl       v3.0.0 v3.0.0 Symfony Security Component - ACL (Access Control List)
symfony/swiftmailer-bundle v2.6.3 v3.1.0 Symfony SwiftmailerBundle
symfony/symfony            v3.3.9 v3.3.9 The Symfony PHP framework
```

```
PHP 7.1.10 (cli) (built: Sep 27 2017 08:27:18) ( NTS )
Copyright (c) 1997-2017 The PHP Group
Zend Engine v3.1.0, Copyright (c) 1998-2017 Zend Technologies
    with Zend OPcache v7.1.10, Copyright (c) 1999-2017, by Zend Technologies
    with Xdebug v2.5.5, Copyright (c) 2002-2017, by Derick Rethans
```

## Subject
I'm trying to make the bundle (more) compatible with Symfony 3 with the help of the work that is already done by @Filoz. I am not sure if this is a compatibility error or some other bug. Any suggestions on how to solve this ?

## Steps to reproduce
1. Install and configure the bundles
2. add a new product
3. press save

## Expected results
Edit form with the new product.

## Actual results
`["validateOneMainCategory"] targeted by Callback constraint is not a valid callable`

```
Symfony\Component\Validator\Exception\ConstraintDefinitionException:
["validateOneMainCategory"] targeted by Callback constraint is not a valid callable

  at vendor/symfony/symfony/src/Symfony/Component/Validator/Constraints/CallbackValidator.php:43
  at Symfony\Component\Validator\Constraints\CallbackValidator->validate(object(Purchase), object(Callback))
     (vendor/symfony/symfony/src/Symfony/Component/Validator/Validator/RecursiveContextualValidator.php:850)
  at Symfony\Component\Validator\Validator\RecursiveContextualValidator->validateInGroup(object(Purchase), '00000000779fc1f8000000006d81b361', object(GenericMetadata), 'Default', object(ExecutionContext))
     (vendor/symfony/symfony/src/Symfony/Component/Validator/Validator/RecursiveContextualValidator.php:696)
  at Symfony\Component\Validator\Validator\RecursiveContextualValidator->validateGenericNode(object(Purchase), null, '00000000779fc1f8000000006d81b361', object(GenericMetadata), 'data', array('Default'), null, 1, object(ExecutionContext))
     (vendor/symfony/symfony/src/Symfony/Component/Validator/Validator/RecursiveContextualValidator.php:141)
  at Symfony\Component\Validator\Validator\RecursiveContextualValidator->validate(object(Purchase), array(object(Callback)), array('Default'))
     (vendor/sonata-project/core-bundle/Validator/ErrorElement.php:224)
  at Sonata\CoreBundle\Validator\ErrorElement->validate(object(Callback))
     (vendor/sonata-project/core-bundle/Validator/ErrorElement.php:99)
  at Sonata\CoreBundle\Validator\ErrorElement->__call('assertCallback', array(array('validateOneMainCategory')))
     (vendor/sonata-project/ecommerce/src/ProductBundle/Admin/ProductAdmin.php:222)
  at Sonata\ProductBundle\Admin\ProductAdmin->validate(object(ErrorElement), object(Purchase))
     (vendor/sonata-project/admin-bundle/Admin/AbstractAdmin.php:3285)
  at Sonata\AdminBundle\Admin\AbstractAdmin->Sonata\AdminBundle\Admin\{closure}(object(ErrorElement), object(Purchase))
  at call_user_func(object(Closure), object(ErrorElement), object(Purchase))
     (vendor/sonata-project/core-bundle/Validator/InlineValidator.php:58)
  at Sonata\CoreBundle\Validator\InlineValidator->validate(object(Purchase), object(InlineConstraint))
     (vendor/symfony/symfony/src/Symfony/Component/Validator/Validator/RecursiveContextualValidator.php:850)
  at Symfony\Component\Validator\Validator\RecursiveContextualValidator->validateInGroup(object(Purchase), '00000000779fc1f8000000006d81b361', object(ClassMetadata), 'Default', object(ExecutionContext))
     (vendor/symfony/symfony/src/Symfony/Component/Validator/Validator/RecursiveContextualValidator.php:563)
  at Symfony\Component\Validator\Validator\RecursiveContextualValidator->validateClassNode(object(Purchase), '00000000779fc1f8000000006d81b361', object(ClassMetadata), 'data', array('Default'), null, 1, object(ExecutionContext))
     (vendor/symfony/symfony/src/Symfony/Component/Validator/Validator/RecursiveContextualValidator.php:368)
  at Symfony\Component\Validator\Validator\RecursiveContextualValidator->validateObject(object(Purchase), 'data', array('Default'), 1, object(ExecutionContext))
     (vendor/symfony/symfony/src/Symfony/Component/Validator/Validator/RecursiveContextualValidator.php:162)
  at Symfony\Component\Validator\Validator\RecursiveContextualValidator->validate(object(Purchase), null, array('Default'))
     (vendor/symfony/symfony/src/Symfony/Component/Form/Extension/Validator/Constraints/FormValidator.php:52)
  at Symfony\Component\Form\Extension\Validator\Constraints\FormValidator->validate(object(Form), object(Form))
     (vendor/symfony/symfony/src/Symfony/Component/Validator/Validator/RecursiveContextualValidator.php:850)
  at Symfony\Component\Validator\Validator\RecursiveContextualValidator->validateInGroup(object(Form), '00000000779fccc4000000006d81b361', object(ClassMetadata), 'Default', object(ExecutionContext))
     (vendor/symfony/symfony/src/Symfony/Component/Validator/Validator/RecursiveContextualValidator.php:563)
  at Symfony\Component\Validator\Validator\RecursiveContextualValidator->validateClassNode(object(Form), '00000000779fccc4000000006d81b361', object(ClassMetadata), '', array('Default'), null, 1, object(ExecutionContext))
     (vendor/symfony/symfony/src/Symfony/Component/Validator/Validator/RecursiveContextualValidator.php:368)
  at Symfony\Component\Validator\Validator\RecursiveContextualValidator->validateObject(object(Form), '', array('Default'), 1, object(ExecutionContext))
     (vendor/symfony/symfony/src/Symfony/Component/Validator/Validator/RecursiveContextualValidator.php:162)
  at Symfony\Component\Validator\Validator\RecursiveContextualValidator->validate(object(Form), null, array('Default'))
     (vendor/symfony/symfony/src/Symfony/Component/Validator/Validator/RecursiveValidator.php:115)
  at Symfony\Component\Validator\Validator\RecursiveValidator->validate(object(Form))
     (vendor/symfony/symfony/src/Symfony/Component/Form/Extension/Validator/EventListener/ValidationListener.php:55)
  at Symfony\Component\Form\Extension\Validator\EventListener\ValidationListener->validateForm(object(FormEvent), 'form.post_bind', object(EventDispatcher))
  at call_user_func(array(object(ValidationListener), 'validateForm'), object(FormEvent), 'form.post_bind', object(EventDispatcher))
     (vendor/symfony/symfony/src/Symfony/Component/EventDispatcher/EventDispatcher.php:212)
  at Symfony\Component\EventDispatcher\EventDispatcher->doDispatch(array(array(object(ValidationListener), 'validateForm'), array(object(DataCollectorListener), 'postSubmit')), 'form.post_bind', object(FormEvent))
     (vendor/symfony/symfony/src/Symfony/Component/EventDispatcher/EventDispatcher.php:44)
  at Symfony\Component\EventDispatcher\EventDispatcher->dispatch('form.post_bind', object(FormEvent))
     (vendor/symfony/symfony/src/Symfony/Component/EventDispatcher/ImmutableEventDispatcher.php:43)
  at Symfony\Component\EventDispatcher\ImmutableEventDispatcher->dispatch('form.post_bind', object(FormEvent))
     (vendor/symfony/symfony/src/Symfony/Component/Form/Form.php:667)
  at Symfony\Component\Form\Form->submit(array(), true)
     (vendor/symfony/symfony/src/Symfony/Component/Form/Extension/HttpFoundation/HttpFoundationRequestHandler.php:113)
  at Symfony\Component\Form\Extension\HttpFoundation\HttpFoundationRequestHandler->handleRequest(object(Form), object(SiteRequest))
     (vendor/symfony/symfony/src/Symfony/Component/Form/Form.php:500)
  at Symfony\Component\Form\Form->handleRequest(object(SiteRequest))
     (vendor/sonata-project/admin-bundle/Controller/CRUDController.php:507)
  at Sonata\AdminBundle\Controller\CRUDController->createAction(object(SiteRequest))
     (vendor/sonata-project/ecommerce/src/ProductBundle/Controller/ProductAdminController.php:43)
  at Sonata\ProductBundle\Controller\ProductAdminController->createAction(object(SiteRequest))
  at call_user_func_array(array(object(ProductAdminController), 'createAction'), array(object(SiteRequest)))
     (vendor/symfony/symfony/src/Symfony/Component/HttpKernel/HttpKernel.php:153)
  at Symfony\Component\HttpKernel\HttpKernel->handleRaw(object(SiteRequest), 1)
     (vendor/symfony/symfony/src/Symfony/Component/HttpKernel/HttpKernel.php:68)
  at Symfony\Component\HttpKernel\HttpKernel->handle(object(SiteRequest), 1, true)
     (vendor/symfony/symfony/src/Symfony/Component/HttpKernel/Kernel.php:171)
  at Symfony\Component\HttpKernel\Kernel->handle(object(SiteRequest))
     (web/app_dev.php:31)

```

### PR DESCRIPTION
**Pull request by pierredup**
## Subject

Fix the validator compatibility with Symfony 3

I am targeting this branch, because it is the only branch with support for Symfony 3.

Closes #470 

## Changelog

```markdown
### Fixed
- validator parameters to be compatible with Symfony 3
```

### CODE REVIEW

**Comment by greg0ire** [comment]
Please write a changelog directed at end users. Keep in mind that they probably do not care what file or class you changed, but what features you provided / what bugs you fixed. Try to answer the question "What benefit do I get from upgrading?"
More on http://keepachangelog.com

File: src/BasketBundle/Controller/BasketController.php [hunk 1]
```diff
@@ -468,7 +468,7 @@ public function finalReviewStepAction(Request $request)
 
         $violations = $this
             ->get('validator')
-            ->validate($basket, ['elements', 'delivery', 'payment']);
+            ->validate($basket, null, ['elements', 'delivery', 'payment']);
```
Comment (greg0ire) [inline src/BasketBundle/Controller/BasketController.php]: If I understand correctly, this method has always been like this: https://github.com/symfony/symfony/commit/718601c6c3642fa7a65a408016a34feef00fe44d#diff-4666f31ca4bcde46b4880e4a8d20e584R49

So you are not only fixing compatibility with Symfony 3, and should probably target the stable branch, at least for this change, correct?

File: src/ProductBundle/Admin/ProductAdmin.php [hunk 2]
```diff
@@ -169,7 +169,7 @@ public function getProduct()
     public function validate(ErrorElement $errorElement, $object): void
     {
         $errorElement
-            ->assertCallback(['validateOneMainCategory'])
+            ->assertCallback(['callback' => 'validateOneMainCategory'])
```
Comment (greg0ire) [inline src/ProductBundle/Admin/ProductAdmin.php]: Same here, are you sure this does not need to go to the stable branch too?

File: src/BasketBundle/Controller/BasketController.php (hunk 1 above)
Comment (pierredup) [reply src/BasketBundle/Controller/BasketController.php]: The validator services uses the `RecursiceValidator` class, which only had a signature change in 3.0 (https://github.com/symfony/symfony/commit/2a6b629e39a3069f67af94e991c7a1b672c7b983#diff-8ca64353ad65881b21945a227035b8e0L118)
In Symfony 2.8, the validator class implements 2 interfaces, the legacy interface with the old signature, and the new interface with the new signature

File: src/ProductBundle/Admin/ProductAdmin.php (hunk 2 above)
Comment (pierredup) [reply src/ProductBundle/Admin/ProductAdmin.php]: In Symfony 2.8, if you didn't specify a `callback` or `methods` key, it added it automatically. This changed in 3.0 where the key is now required. So although this change will work on Symfony 2.8 as well, it is not required so I think it's fine to only target the dev branch

File: src/BasketBundle/Controller/BasketController.php (hunk 1 above)
Comment (greg0ire) [reply src/BasketBundle/Controller/BasketController.php]: Oh ok then…

**Comment by greg0ire** [comment]
Thanks @pierredup !

**Pull request merged**